src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,905 @@
+	long offset_offset;
+	/** Offset of psize field in the file header. */
+	long psize_offset;
+	/** Offset of len field in the file header. */
+	long len_offset;
+	/** The data is fully written into the file already. */
+	bool is_committed;
+};
+
+#define mh_name _memtx_sort_data_entries
+#define mh_key_t struct memtx_sort_data_key
+#define mh_node_t struct memtx_sort_data_reader_entry
+#define mh_arg_t void *
+#define mh_hash(a, _) (memtx_sort_data_key_hash((a)->key))
+#define mh_hash_key(a, _) (memtx_sort_data_key_hash(a))
+#define mh_cmp(a, b, _) (memtx_sort_data_key_cmp((a)->key, (b)->key))
+#define mh_cmp_key(a, b, _) (memtx_sort_data_key_cmp(a, (b)->key))
+#define MH_SOURCE
+#include "salad/mhash.h"
+
+#define mh_name _memtx_sort_data_writer_entries
+#define mh_key_t struct memtx_sort_data_key
+#define mh_node_t struct memtx_sort_data_writer_entry
+#define mh_arg_t void *
+#define mh_hash(a, _) (memtx_sort_data_key_hash((a)->key))
+#define mh_hash_key(a, _) (memtx_sort_data_key_hash(a))
+#define mh_cmp(a, b, _) (memtx_sort_data_key_cmp((a)->key, (b)->key))
+#define mh_cmp_key(a, b, _) (memtx_sort_data_key_cmp(a, (b)->key))
+#define MH_SOURCE
+#include "salad/mhash.h"
+
+/** The sort data file writer context. */
+struct memtx_sort_data_writer {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The sort data file name. */
+	char fname[PATH_MAX];
+	/** The information about the sort data entries. */
+	struct mh_memtx_sort_data_writer_entries_t *entries;
+	/** The currently handled sort data entry. */
+	struct memtx_sort_data_writer_entry *curr_entry;
+	/** The total amount of tuples in saved spaces. */
+	size_t cardinality;
+	/** The offset to the "Cardinality" file header key. */
+	long cardinality_offset;
+};
+
+/** The sort data reader context. */
+struct memtx_sort_data_reader {
+	/** The sort data file pointer. */
+	FILE *fp;
+	/** The sort data file name. */
+	char fname[PATH_MAX];
+	/** The information about the sort data entries. */
+	struct mh_memtx_sort_data_entries_t *entries;
+	/** The currently handled sort data entry, NULL if space is skipped. */
+	struct memtx_sort_data_reader_entry *curr_entry;
+	/** The currently handled space ID. */
+	uint32_t curr_space_id;
+	/** The old to new tuple address map used on recovery. */
+	struct mh_ptrptr_t *old2new;
+	/** The buffer to use on fread. */
+	void *buffer;
+};
+
+/* {{{ Writer *****************************************************************/
+
+struct memtx_sort_data_writer *
+memtx_sort_data_writer_new(struct read_view *rv)
+{
+	struct memtx_sort_data_writer *writer = xmalloc(sizeof(*writer));
+	writer->fp = NULL;
+	memset(writer->fname, 0, sizeof(writer->fname));
+	writer->entries = mh_memtx_sort_data_writer_entries_new();
+	writer->curr_entry = NULL;
+	writer->cardinality = 0;
+	writer->cardinality_offset = 0;
+
+	/* Find out which spaces and indexes must be saved. */
+	struct space_read_view *space_rv;
+	read_view_foreach_space(space_rv, rv) {
```

Comment:
The sort data writer API should be designed in such a way that it doesn't require a walk over all indexes in a read view. This should be done only by the checkpoint writer.

Reply:
Done, now the filtering happens during the read view creation. There's a problem though.

The sort data file format implies we know the amount of saved indexes prior to start of write: we have a header with IDs and offsets/sizesm and then we have the sort data. So we have to traverse the dumped indexes prior to data dump to calculate the amount of indexes to be dumped.